pkg/controller/jobframework/reconciler.go

```diff
@@ -333,8 +346,12 @@ func (r *JobReconciler) stopJob(ctx context.Context, job GenericJob, object clie
 		}
 	}
 
-	if wl != nil {
-		job.RestoreNodeAffinity(wl.Spec.PodSets)
+	log.V(3).Info("restore node selectors from annotation")
+	selectors, err := getNodeSelectorsFromObjectAnnotation(object)
+	if err != nil {
+		log.V(3).Error(err, "Unable to get original node selectors")
+	} else {
+		job.RestoreNodeAffinity(selectors)
```

Comment:
we could fallback into getting the selectors from the workload object

Reply:
It was my original approach but was change during the review.

Reply:
IIRC in the original approach you first looked up the workload object and then fallbacked to annotation. 

Anyway, I think that once we made the annotation immutable, we can fully rely on it, unless I'm missing some scenario. 

If such a scenario exists my point was that we should add a comment why this is done. Otherwise we will end up in a suspicious code which no-one remembers / knows why needed.

Reply:
The order was inverted due to performance reasons, getting the selectors from the workload  wold not have needed additional un-marshaling.

The scenario wold be when a job is missing the annotation,  could happen if the job (other than core.Job or MPIJob) in question is not blocking the change of the annotation while running.

Reply:
Yes, but we have a webhookbto prevent the change. So, the only scenario is that the webhook is not functioning for some reason and we have a bad actor. However, if the webhook is malfunctioning and we have a bad actor, the actor could both modify the annotation and delete the workload sk the fallback would not work either. So, iiuc, it would not be bullet proof either, just misleadingly making that impression.